src/vcpkg/base/git.cpp

```diff
@@ -11,288 +10,498 @@
+                                                  msg::command_line = maybe_ls_tree_result.command.command_line());
+                    result.clear();
+                    return result;
                 }
+
+                ret.push_back(GitLSTreeEntry{split_line[1], file_info_section.back()});
             }
 
-            if (!ParserBase::is_lineend(parser.cur()))
+            return result;
+        }
+
+        result.clear();
+        return result;
+    }
+
+    bool git_read_tree(DiagnosticContext& context,
+                       const Filesystem& fs,
+                       const Path& git_exe,
+                       GitRepoLocator locator,
+                       const Path& destination,
+                       StringView treeish)
+    {
+        auto pid = get_process_id();
+        Path git_tree_temp = fmt::format("{}_{}.tmp", destination, pid);
+        git_tree_temp.make_generic();
+        Path git_tree_index = fmt::format("{}_{}.index", destination, pid);
+        auto parent = destination.parent_path();
+        if (!parent.empty())
+        {
+            if (!fs.create_directories(context, parent).has_value())
             {
-                parser.add_error(msg::format(msgGitStatusOutputExpectedNewLine), parser.cur_loc());
-                break;
+                return false;
             }
+        }
+
+        if (!fs.remove_all(context, git_tree_temp) || !fs.create_directory(context, git_tree_temp).has_value())
+        {
+            return false;
+        }
 
-            parser.next();
-            results.push_back(result);
+        StringView read_tree_args[] = {StringLiteral{"read-tree"}, treeish};
+        auto maybe_git_read_tree_output =
+            run_git_cmd_with_index(context, git_exe, locator, git_tree_index, read_tree_args);
+        if (maybe_git_read_tree_output.maybe_output.has_value())
+        {
+            auto prefix_arg = fmt::format("--prefix={}/", git_tree_temp);
+            StringView checkout_index_args[] = {StringLiteral{"--work-tree"},
+                                                git_tree_temp,
+                                                StringLiteral{"checkout-index"},
+                                                StringLiteral{"-af"},
+                                                StringLiteral{"--ignore-skip-worktree-bits"},
+                                                prefix_arg};
+            auto git_checkout_index_command =
+                run_git_cmd_with_index(context, git_exe, locator, git_tree_index, checkout_index_args);
+            bool succeeded = git_checkout_index_command.maybe_output.has_value();
+            fs.remove(git_tree_index, IgnoreErrors{});
+            return succeeded && fs.rename_or_delete(context, git_tree_temp, destination).has_value();
         }
 
-        if (auto error = parser.get_error())
+        if (is_shallow_clone(context, git_exe, locator).value_or(false))
```

Comment:
If this fails, will the user will see a weird, unrelated error to their actual problem?

Maybe information gathering shouldn't report errors?